src/Database/Database.php

```diff
@@ -2919,15 +2919,113 @@ public function updateDocument(string $collection, string $id, Document $documen
         if ($collection->getId() !== self::METADATA) {
             $documentSecurity = $collection->getAttribute('documentSecurity', false);
 
-            $relationshipKeys = [];
             foreach ($relationships as $relationship) {
                 $relationshipKey = $relationship->getAttribute('key');
-                $relationshipKeys[$relationshipKey] = $relationshipKey;
+                $relationships[$relationshipKey] = $relationship;
             }
             // Compare if the document has any changes
             foreach ($document as $key=>$value) {
                 // Skip the nested documents as they will be checked later in recursions.
-                if (array_key_exists($key, $relationshipKeys)) {
+                if (array_key_exists($key, $relationships)) {
+                    $relationType = (string) $relationships[$key]['options']['relationType'];
+                    $side = (string) $relationships[$key]['options']['side'];
+                    switch($relationType) {
+                        case Database::RELATION_ONE_TO_ONE: {
+                            $oldValue = $old->getAttribute($key) instanceof Document ? $old->getAttribute($key)->getId() : $old->getAttribute($key);
+                            if (
+                                (\is_null($old->getAttribute($key)) && !\is_null($value)) ||
+                                (!is_null($old->getAttribute($key)) && \is_null($value))) {
+                                $shouldUpdate = true;
+                                break;
+                            } elseif (\is_string($value)) {
+                                if ($oldValue !== $value) {
+                                    $shouldUpdate = true;
+                                    break;
+                                }
+                            } elseif ($oldValue !== $value->getId()) {
+                                $shouldUpdate = true;
+                                break;
+                            }
```

Comment:
We can simplify this:

```suggestion
                            if (
                                (\is_null($value) !== \is_null($oldValue)) ||
                                (\is_string($value) && $oldValue !== $value) ||
                                ($oldValue !== $value->getId())
                            ) {
                                $shouldUpdate = true;
                                break;
                            }
```

Reply:
updated the if conditions